monkey/agent_plugins/exploiters/smb/src/smb_exploiter.py

```diff
@@ -1,305 +1,102 @@
-        if smb.isGuestSession() > 0:
-            try:
-                smb.logoff()
-            except Exception:
-                # TODO: If we failed to logout, we should handle that
-                pass
+    def _exploit(
+        self, host: TargetHost, options: SMBOptions, interrupt: Event
+    ) -> Optional[Credentials]:
+        propagated_credentials = self._brute_force(host, options, interrupt)
 
-            return True
-        return False
+        return propagated_credentials
 
-    @classmethod
-    def query_server_info(self, smb: SMBConnection):
-        try:
-            info = SMB.execute_rpc_call(smb, "hNetrServerGetInfo", 102)
-        except Exception as err:
-            logger.debug(f"Failed to query server info: {err}")
-            return None
-
-        return info
-
-    @classmethod
-    def query_shared_resources(self, smb: SMBConnection):
-        try:
-            shares = SMB.execute_rpc_call(smb, "hNetrShareEnum", 2)
-        except Exception as err:
-            logger.debug(f"Failed to query shared resources: {err}")
-            return None
+    def _brute_force(
+        self, host: TargetHost, options: SMBOptions, interrupt: Event
+    ) -> Optional[Credentials]:
+        credentials = self._propagation_credentials_repository.get_credentials()
+        credentials_list = generate_brute_force_credentials(credentials)
 
-        return shares
+        for brute_force_credentials in interruptible_iter(credentials_list, interrupt):
+            if self._exploit_client.authenticate(host, options, brute_force_credentials):
+                return brute_force_credentials
 
-    @staticmethod
-    def execute_rpc_call(smb, rpc_func, *args):
-        dce = SMB.get_dce_bind(smb)
-        rpc_method_wrapper = getattr(srvs, rpc_func, None)
-        if not rpc_method_wrapper:
-            raise ValueError("Cannot find RPC method '%s'" % (rpc_method_wrapper,))
-
-        return rpc_method_wrapper(dce, *args)
+        return None
 
-    @staticmethod
-    def get_dce_bind(smb: SMBConnection) -> DCERPC_v5:
-        rpctransport = transport.SMBTransport(
-            smb.getRemoteHost(), smb.getRemoteHost(), filename=r"\srvsvc", smb_connection=smb
+    def _propagate(
+        self,
+        host: TargetHost,
+        options: SMBOptions,
+        credentials: Credentials,
+        servers: Sequence[str],
+        current_depth: int,
+        interrupt: Event,
+    ) -> bool:
+        destination_path = get_agent_dst_path(host)
+        agent_binary = self._agent_binary_repository.get_agent_binary(host.operating_system)
+        remote_path = self._propagation_client.copy_file(
+            agent_binary, host, destination_path, credentials
         )
-        dce = rpctransport.get_dce_rpc()
-        dce.connect()
-        dce.bind(srvs.MSRPC_UUID_SRVS)
+        if not remote_path:
+            logger.debug(f"Failed to copy agent to {host}")
+            return False
```

Comment:
This is fine for SMB, but not necessarily generalizable for other brute force exploiters. Why not have `copy_file()` raise some kind of `RemoteFileCopyError` if copy fails?